workflow/controller/operator.go

```diff
@@ -3199,14 +3178,45 @@ func (woc *wfOperationCtx) setExecWorkflow() error {
 		woc.execWf = &wfv1.Workflow{Spec: *woc.wf.Status.StoredWorkflowSpec.DeepCopy()}
 		woc.volumes = woc.execWf.Spec.DeepCopy().Volumes
 	} else if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
-		return fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+		err := fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+		woc.log.WithError(err).Errorf("Unable to get Workflow Template Reference for workflow")
+		woc.markWorkflowError(ctx, err)
+		return err
 	} else {
 		err := woc.controller.setWorkflowDefaults(woc.wf)
 		if err != nil {
+			woc.log.WithError(err).Errorf("Unable to get Workflow Template Reference for workflow")
+			woc.markWorkflowError(ctx, err)
 			return err
 		}
 		woc.volumes = woc.wf.Spec.DeepCopy().Volumes
 	}
+
+	// Perform one-time workflow validation
+	if woc.wf.Status.Phase == wfv1.WorkflowUnknown {
+		validateOpts := validate.ValidateOpts{ContainerRuntimeExecutor: woc.getContainerRuntimeExecutor()}
+		wftmplGetter := templateresolution.WrapWorkflowTemplateInterface(woc.controller.wfclientset.ArgoprojV1alpha1().WorkflowTemplates(woc.wf.Namespace))
+		cwftmplGetter := templateresolution.WrapClusterWorkflowTemplateInterface(woc.controller.wfclientset.ArgoprojV1alpha1().ClusterWorkflowTemplates())
+
+		// Validate the execution wfSpec
+		wfConditions, err := validate.ValidateWorkflow(wftmplGetter, cwftmplGetter, woc.wf, validateOpts)
+		if err != nil {
+			msg := fmt.Sprintf("invalid spec: %s", err.Error())
+			woc.markWorkflowFailed(ctx, msg)
+			return err
+		}
+
+		// If we received conditions during validation (such as SpecWarnings), add them to the Workflow object
+		if len(*wfConditions) > 0 {
+			woc.wf.Status.Conditions.JoinConditions(wfConditions)
+			woc.updated = true
+		}
+	}
+	woc.setGlobalParameters(woc.execWf.Spec.Arguments)
+	err := woc.substituteGlobalVariables()
```

Comment:
Should these two be the same function?

Reply:
No, One function is populating a variable in the globalParammap to substitute the value in all places like here and executeTemplate. 
The second one is a new function that will substitute the global variable values into WorkflowSpec.